{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import re"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "#make a class for storing data and functions for GTF files\n",
    "\n",
    "class GTF_Parser: \n",
    "    \"\"\"\n",
    "    For gtf files\n",
    "    \"\"\"\n",
    "\n",
    "    def __init__(self, file_name):\n",
    "        \"\"\"\n",
    "        Initialize self.\n",
    "        And create a dataframe from the gtf file (self) data, with separate columns for gene ID and gene names. \n",
    "        \"\"\"\n",
    "        \n",
    "        self.df=pd.read_table(file_name, sep='\\t', skiprows=1, header=None)\n",
    "        \n",
    "        #Give names to the columns and create two additional columns for Gene name and ID\n",
    "        self.df.columns=['seqname','source','feature','start','end','score','strand','frame','attribute']\n",
    "        self.df['Gene_ID']=\"\"\n",
    "        self.df['Gene_name']=\"\"\n",
    "        \n",
    "        self.attributes=self.df['attribute']\n",
    "        \n",
    "        #I will extract the name and ID of each gene from the 'attribute' column and add them to separate columns\n",
    "        \n",
    "        for index, value in self.attributes.items():\n",
    "            separate=value.split(\";\")                #the ID and name are spearated by ';' in the attribute column\n",
    "            \n",
    "            for feature in separate:\n",
    "                \n",
    "                if re.search(r\"gene_name\",feature):\n",
    "                    feat_name=feature.split('\"')\n",
    "                    self.df['Gene_name'][index]=feat_name[1]\n",
    "                    \n",
    "                if re.search(r\"gene_id\",feature):\n",
    "                    feat_id=feature.split('\"')\n",
    "                    self.df['Gene_ID'][index]=feat_id[1]\n",
    "                    \n",
    "        #If some entries have only ID and no name, then the 'Gene name' column is filled with the ID (and vice versa):\n",
    "                    \n",
    "        for index, value in self.df['Gene_ID'].items():\n",
    "            if value == '' :\n",
    "                self.df['Gene_ID'][index]=self.df['Gene_name'][index]\n",
    "                \n",
    "        for index, value in self.df['Gene_name'].items():\n",
    "            if value == '' :\n",
    "                self.df['Gene_name'][index]=self.df['Gene_ID'][index]\n",
    "                \n",
    "                \n",
    "\n",
    "    def printCoordinates(self, gene_name):\n",
    "        \"\"\"\n",
    "        Takes a gene name as an argument.\n",
    "        For a protein coding gene:\n",
    "            ~if there are 3 features for the desired gene name, it finds and prints the start and end after joining the 3 features (They should be 2 UTRs and 1 CDS).\n",
    "            ~if there are 1, 2 or more than 3 features for the desired gene name, it finds and prints the start and end for each feature separately, indicating what kind of feature it is (should be either UTR or CDS)\n",
    "        For a non protein-coding gene:\n",
    "            it finds and prints start and end, indicating what kind of gene it is (e.g. sRNA, intergenic region)\n",
    "        If the gene name doesn't exist in 'Gene_name' column, it looks for it in the 'Gene_ID' column.\n",
    "        If the gene name doesn't exist neither in 'Gene_name' nor in 'Gene_ID' columns, it returns nothing.\n",
    "        \"\"\"\n",
    "        \n",
    "        start_positions=[]\n",
    "        end_positions=[]\n",
    "        prot_indexes=[]\n",
    "        \n",
    "        for index, value in self.df['Gene_name'].items():\n",
    "            \n",
    "            #It will look for the gene name in the 'Gene name' and 'Gene ID' columns:\n",
    "            \n",
    "            if re.search(gene_name, value) or re.search(gene_name, self.df['Gene_ID'][index]):\n",
    "                \n",
    "                #If the gene is protein coding, I am saving the coordinates of all the features in lists, so that I can merge \n",
    "                #later the untranslated regions with the coding sequence and return the coordinates of the total region:\n",
    "                \n",
    "                if self.df['source'][index]=='protein_coding':\n",
    "                    start_positions.append(self.df['start'][index])\n",
    "                    end_positions.append(self.df['end'][index])\n",
    "                    prot_indexes.append(index)\n",
    "                    \n",
    "                else:\n",
    "                    print (self.df['source'][index]+ \" \" +gene_name+ \" starts at \"+ str(self.df['start'][index]) +\" and ends at \"+str(self.df['end'][index]))\n",
    "        \n",
    "        starts=sorted(start_positions)\n",
    "        ends=sorted(end_positions)\n",
    "        \n",
    "        if len(starts) != 0:\n",
    "            \n",
    "            if len(starts) == 3:\n",
    "                start_point=starts[0]\n",
    "                end_point=ends[-1]\n",
    "                print (\"Protein coding gene \"+gene_name+\" starts at \"+str(start_point)+\" and ends at \"+str(end_point))\n",
    "                \n",
    "            #If the gene is protein coding, but for some reason has more (or less) features than 3 (5'UTR, CDS, 3'UTR),\n",
    "            #the coordinates of each separate feature will be printed:\n",
    "            \n",
    "            else:\n",
    "                print (\"There are \"+str(len(starts))+\" protein coding features with name \"+gene_name)\n",
    "                for i in prot_indexes:\n",
    "                    print (\"Protein coding \"+self.df['feature'][i]+\" feature starts at \"+self.df['start'][i]+\" and ends at \"+self.df['end'][i])\n",
    "        \n",
    "        \n",
    "    \n",
    "    def getCoordinates(self, gene_name): \n",
    "        \"\"\"\n",
    "            Takes a gene name as an argument.\n",
    "            -For a protein coding gene:\n",
    "                 it finds and prints the start and end after joining the 5UTR, CDS and 3UTR. \n",
    "            -For a non protein-coding gene:\n",
    "                it finds and prints start and end (if there are multiple entries for the same name, it joins them all first)\n",
    "            \"\"\"\n",
    "\n",
    "        #First, we create a dataframe containing only the rows related to our desired gene\n",
    "\n",
    "        condition1=self.df['Gene_name'].str.contains(gene_name)\n",
    "        general_df=self.df.loc[condition1]   \n",
    "\n",
    "        #Then, we create another dataframe containing only the 'protein coding' rows related to our desired gene\n",
    "                #If the gene is not protein coding, then this new dataframe is empty (has 0 rows)\n",
    "\n",
    "        condition2=self.df['source']=='protein_coding'\n",
    "        protcod_df=self.df.loc[condition1&condition2]    \n",
    "\n",
    "\n",
    "        #if the gene cannot be found (then both dataframes will have 0 rows):\n",
    "        if (protcod_df.shape[0] == 0) & (general_df.shape[0] == 0):\n",
    "\n",
    "            pass\n",
    "\n",
    "        #if the protcod dataframe is empty (number of rows=0_, which means that the gene is not protein coding):\n",
    "        elif protcod_df.shape[0] == 0:    \n",
    "\n",
    "            start_positions=[]\n",
    "            end_positions=[]\n",
    "            general_df.apply(lambda x: start_positions.append(x['start']), axis=1)\n",
    "            general_df.apply(lambda x: end_positions.append(x['end']), axis=1)\n",
    "                    #this is a bit redundant, as if we look for a tRNA or sRNA, our dataframe will probably have only 1 entry\n",
    "            return [start_positions[0], end_positions[-1]]\n",
    "\n",
    "        #if the gene is protein coding:\n",
    "        else:\n",
    "            start_positions=[]\n",
    "            end_positions=[]\n",
    "            protcod_df.apply(lambda x: start_positions.append(x['start']), axis=1)\n",
    "            protcod_df.apply(lambda x: end_positions.append(x['end']), axis=1)\n",
    "            starts=sorted(start_positions)\n",
    "            ends=sorted(end_positions)\n",
    "            start=starts[0]\n",
    "            end=ends[-1]\n",
    "\n",
    "            return [start_positions[0], end_positions[-1]]\n",
    "\n",
    "    ## Even if the code in the 'if' is basically the same both for coding and non-coding genes, the difference is that we use \n",
    "    ## different dataframes, so that we get exclude any intergenic regions when we have a protein coding gene         \n",
    "\n",
    "    \n",
    "    \n",
    "    def getCDScoordinates(self, gene_name):\n",
    "        \"\"\"\n",
    "        Takes a gene name as argument. \n",
    "        It returns the start and end of the gene's coding sequence. \n",
    "        If the gene name doesn't exist or it is not proteing coding gene, it does nothing.\n",
    "        \"\"\"\n",
    "        \n",
    "        #First, we create a dataframe containing only the CDS row of our desired gene\n",
    "        #(if we give a protein coding gene name, the new_df will have 1 row, otherwise it will have 0)\n",
    "\n",
    "        condition=(self.df['Gene_name'].str.contains(gene_name)) & (self.df['feature']=='CDS')\n",
    "        new_df=self.df.loc[condition]   \n",
    "        \n",
    "        #The 'if' statement below is added in order not to take an error when we give a non-coding gene name \n",
    "        \n",
    "        if new_df.shape[0] != 0:       #if the number of rows is not 0\n",
    "            \n",
    "            start=new_df.iloc[0].loc['start']     #.iloc[0]: 1st row   #.loc['start']: start column\n",
    "            end=new_df.iloc[0].loc['end']\n",
    "\n",
    "            return [start, end]\n",
    "    \n",
    "    \n",
    "    \n",
    "    def getCDSLength(self, gene_name):\n",
    "        \"\"\"\n",
    "        Takes a gene name as an argument. \n",
    "        Returns the length of its CDS sequence. \n",
    "        If the gene name doesn't exist or it is not proteing coding gene, it returns nothing.\n",
    "        \"\"\"\n",
    "        \n",
    "        #We create a dataframe containing only the CDS row of our desired gene. We do that only for not getting an error\n",
    "        #in case we give a noncoding gene\n",
    "        condition=(self.df['Gene_name'].str.contains(gene_name)) & (self.df['feature']=='CDS')\n",
    "        new_df=self.df.loc[condition] \n",
    "        \n",
    "        if new_df.shape[0] !=0:          #if the number of rows is not 0, meaning if the gene is protein coding\n",
    "        \n",
    "            length=self.getCDScoordinates(gene_name)[1]-self.getCDScoordinates(gene_name)[0]\n",
    "        \n",
    "            return (length)\n",
    "    \n",
    "    \n",
    "    \n",
    "    def getCDSsequence(self, gene_name, genome_file):\n",
    "        \"\"\"\n",
    "        Takes as arguments a gene name and the name of a file containing the genome in fasta format.\n",
    "        Returns the coding sequence (CDS) of the gene, extracted from the genome fasta file, based on its coordinates in the gtf.\n",
    "        \"\"\"\n",
    "        \n",
    "        #Fisrt, we make a variable containing the full genome sequence:\n",
    "        \n",
    "        gnm_file=open(genome_file)\n",
    "        genome_seq=''\n",
    "        for line in gnm_file:\n",
    "            if line[0]!='>':\n",
    "                line=line.rstrip('\\n')\n",
    "                genome_seq=genome_seq+line\n",
    "        \n",
    "        #Now we extract the gene CDS sequence from the full genome sequence, only if the gene is protein coding:\n",
    "        \n",
    "        condition=(self.df['Gene_name'].str.contains(gene_name)) & (self.df['feature']=='CDS')\n",
    "        new_df=self.df.loc[condition] \n",
    "        \n",
    "        if new_df.shape[0] !=0:          #if the number of rows is not 0, meaning if the gene is protein coding\n",
    "        \n",
    "            start=self.getCDScoordinates(gene_name)[0]-1       #Python is 0-based\n",
    "            end=self.getCDScoordinates(gene_name)[1]           #the end is exclusive, so I won't subtract 1\n",
    "            sequence=genome_seq[start:end]\n",
    "            \n",
    "            return (sequence)\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "<ipython-input-3-5daa3b2122f5>:32: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self.df['Gene_name'][index]=feat_name[1]\n",
      "<ipython-input-3-5daa3b2122f5>:36: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self.df['Gene_ID'][index]=feat_id[1]\n",
      "<ipython-input-3-5daa3b2122f5>:42: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  self.df['Gene_ID'][index]=self.df['Gene_name'][index]\n"
     ]
    }
   ],
   "source": [
    "staph_annot=GTF_Parser(\"Staphylococcus_aureus_usa300_FPR3757_1.3_with_intergenic.gtf\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>seqname</th>\n",
       "      <th>source</th>\n",
       "      <th>feature</th>\n",
       "      <th>start</th>\n",
       "      <th>end</th>\n",
       "      <th>score</th>\n",
       "      <th>strand</th>\n",
       "      <th>frame</th>\n",
       "      <th>attribute</th>\n",
       "      <th>Gene_ID</th>\n",
       "      <th>Gene_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CP000255.1</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>exon</td>\n",
       "      <td>1</td>\n",
       "      <td>339</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_name \"INT_dnaA\";</td>\n",
       "      <td>INT_dnaA</td>\n",
       "      <td>INT_dnaA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CP000255.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>5UTR</td>\n",
       "      <td>340</td>\n",
       "      <td>544</td>\n",
       "      <td>108</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"SAUSA300_0001\"; gene_name \"dnaA\";</td>\n",
       "      <td>SAUSA300_0001</td>\n",
       "      <td>dnaA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CP000255.1</td>\n",
       "      <td>sRNA</td>\n",
       "      <td>exon</td>\n",
       "      <td>342</td>\n",
       "      <td>539</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>0</td>\n",
       "      <td>gene_id \"srn_0010_sRNA412\"; gene_name \"srn_001...</td>\n",
       "      <td>srn_0010_sRNA412</td>\n",
       "      <td>srn_0010_sRNA412</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CP000255.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CDS</td>\n",
       "      <td>545</td>\n",
       "      <td>1905</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"SAUSA300_0001\"; gene_name \"dnaA\";</td>\n",
       "      <td>SAUSA300_0001</td>\n",
       "      <td>dnaA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CP000255.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>1906</td>\n",
       "      <td>2029</td>\n",
       "      <td>108</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"SAUSA300_0001\"; gene_name \"dnaA\";</td>\n",
       "      <td>SAUSA300_0001</td>\n",
       "      <td>dnaA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7230</th>\n",
       "      <td>CP000255.1</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>exon</td>\n",
       "      <td>2872415</td>\n",
       "      <td>2872417</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_name \"INT_rnpA_rpmH\";</td>\n",
       "      <td>INT_rnpA_rpmH</td>\n",
       "      <td>INT_rnpA_rpmH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7231</th>\n",
       "      <td>CP000255.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>3UTR</td>\n",
       "      <td>2872418</td>\n",
       "      <td>2872445</td>\n",
       "      <td>1536</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"SAUSA300_2648\"; gene_name \"rpmH\";</td>\n",
       "      <td>SAUSA300_2648</td>\n",
       "      <td>rpmH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7232</th>\n",
       "      <td>CP000255.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>CDS</td>\n",
       "      <td>2872446</td>\n",
       "      <td>2872582</td>\n",
       "      <td>.</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"SAUSA300_2648\"; gene_name \"rpmH\";</td>\n",
       "      <td>SAUSA300_2648</td>\n",
       "      <td>rpmH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7233</th>\n",
       "      <td>CP000255.1</td>\n",
       "      <td>protein_coding</td>\n",
       "      <td>5UTR</td>\n",
       "      <td>2872583</td>\n",
       "      <td>2872729</td>\n",
       "      <td>1536</td>\n",
       "      <td>-</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_id \"SAUSA300_2648\"; gene_name \"rpmH\";</td>\n",
       "      <td>SAUSA300_2648</td>\n",
       "      <td>rpmH</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7234</th>\n",
       "      <td>CP000255.1</td>\n",
       "      <td>intergenic</td>\n",
       "      <td>exon</td>\n",
       "      <td>2872730</td>\n",
       "      <td>2872769</td>\n",
       "      <td>.</td>\n",
       "      <td>+</td>\n",
       "      <td>.</td>\n",
       "      <td>gene_name \"INT_rpmH\";</td>\n",
       "      <td>INT_rpmH</td>\n",
       "      <td>INT_rpmH</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>7235 rows × 11 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "         seqname          source feature    start      end score strand frame  \\\n",
       "0     CP000255.1      intergenic    exon        1      339     .      +     .   \n",
       "1     CP000255.1  protein_coding    5UTR      340      544   108      +     .   \n",
       "2     CP000255.1            sRNA    exon      342      539     .      +     0   \n",
       "3     CP000255.1  protein_coding     CDS      545     1905     .      +     .   \n",
       "4     CP000255.1  protein_coding    3UTR     1906     2029   108      +     .   \n",
       "...          ...             ...     ...      ...      ...   ...    ...   ...   \n",
       "7230  CP000255.1      intergenic    exon  2872415  2872417     .      +     .   \n",
       "7231  CP000255.1  protein_coding    3UTR  2872418  2872445  1536      -     .   \n",
       "7232  CP000255.1  protein_coding     CDS  2872446  2872582     .      -     .   \n",
       "7233  CP000255.1  protein_coding    5UTR  2872583  2872729  1536      -     .   \n",
       "7234  CP000255.1      intergenic    exon  2872730  2872769     .      +     .   \n",
       "\n",
       "                                              attribute           Gene_ID  \\\n",
       "0                                gene_name \"INT_dnaA\";           INT_dnaA   \n",
       "1           gene_id \"SAUSA300_0001\"; gene_name \"dnaA\";      SAUSA300_0001   \n",
       "2     gene_id \"srn_0010_sRNA412\"; gene_name \"srn_001...  srn_0010_sRNA412   \n",
       "3           gene_id \"SAUSA300_0001\"; gene_name \"dnaA\";      SAUSA300_0001   \n",
       "4           gene_id \"SAUSA300_0001\"; gene_name \"dnaA\";      SAUSA300_0001   \n",
       "...                                                 ...               ...   \n",
       "7230                        gene_name \"INT_rnpA_rpmH\";      INT_rnpA_rpmH   \n",
       "7231        gene_id \"SAUSA300_2648\"; gene_name \"rpmH\";      SAUSA300_2648   \n",
       "7232         gene_id \"SAUSA300_2648\"; gene_name \"rpmH\";     SAUSA300_2648   \n",
       "7233        gene_id \"SAUSA300_2648\"; gene_name \"rpmH\";      SAUSA300_2648   \n",
       "7234                             gene_name \"INT_rpmH\";           INT_rpmH   \n",
       "\n",
       "             Gene_name  \n",
       "0             INT_dnaA  \n",
       "1                 dnaA  \n",
       "2     srn_0010_sRNA412  \n",
       "3                 dnaA  \n",
       "4                 dnaA  \n",
       "...                ...  \n",
       "7230     INT_rnpA_rpmH  \n",
       "7231              rpmH  \n",
       "7232              rpmH  \n",
       "7233              rpmH  \n",
       "7234          INT_rpmH  \n",
       "\n",
       "[7235 rows x 11 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "staph_annot.df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "intergenic rnpA starts at 2872415 and ends at 2872417\n",
      "Protein coding gene rnpA starts at 2871834 and ends at 2872414\n"
     ]
    }
   ],
   "source": [
    "staph_annot.printCoordinates('rnpA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "staph_annot.printCoordinates('rnbA')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "(there is no gene 'rnbA', so it prints nothing)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[2871834, 2872414]"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "staph_annot.getCoordinates('rnpA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[2871972, 2872324]"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "staph_annot.getCDScoordinates('rnpA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "352"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "staph_annot.getCDSLength('rnpA')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'tacttaatctttttattaaaaactttggcaattttaagtacgtgctcaagactattctgtatttgtaaagtcgtcatatctttagctggctgtcttgctattacaataatatctttggccaatatatgcgacttatgtactttgaaattttcacgtattgctcttttaatcttgtttcttaacactgcattacctagttttttagaaacactaatacctaagcgaaaatggtctatttctttattattacaagtgtatacaacaaattgtctgttggctacagaatgaccttttttatatattctctgaaaatctgcattctttttaattcggtaagctttttccaataacat'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "staph_annot.getCDSsequence('rnpA','complete-genome-usa300_fpr3757-nc_007793.fasta')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.2"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
